tokio/src/sync/broadcast.rs

```diff
@@ -688,28 +705,48 @@ impl<T> Receiver<T> {
         if guard.pos() != self.next {
             let pos = guard.pos();
 
-            guard.drop_no_rem_dec();
-
+            // The receiver has read all current values in the channel
             if pos.wrapping_add(self.shared.buffer.len() as u64) == self.next {
+                guard.drop_no_rem_dec();
                 return Err(TryRecvError::Empty);
-            } else {
-                let tail = self.shared.tail.lock().unwrap();
+            }
 
-                // `tail.pos` points to the slot the **next** send writes to.
-                // Because a receiver is lagging, this slot also holds the
-                // oldest value. To make the positions match, we subtract the
-                // capacity.
-                let next = tail.pos.wrapping_sub(self.shared.buffer.len() as u64);
-                let missed = next.wrapping_sub(self.next);
+            let tail = self.shared.tail.lock().unwrap();
 
-                self.next = next;
+            // `tail.pos` points to the slot that the **next** send writes to. If
+            // the channel is closed, the previous slot is the oldest value.
+            let mut adjust = 0;
+            if tail.closed {
+                adjust = 1
+            }
+            let next = tail
+                .pos
+                .wrapping_sub(self.shared.buffer.len() as u64 + adjust);
+
+            let missed = next.wrapping_sub(self.next);
+
+            drop(tail);
 
-                return Err(TryRecvError::Lagged(missed));
+            // The receiver is slow but no values have been missed
+            if missed == 0 {
+                self.next = self.next.wrapping_add(1);
+                return Ok(guard);
             }
+
+            guard.drop_no_rem_dec();
+            self.next = next;
+
+            return Err(TryRecvError::Lagged(missed));
         }
 
         self.next = self.next.wrapping_add(1);
 
+        // If the `CLOSED` bit it set on the slot, the channel is closed
+        if slot.lock.load(SeqCst) & CLOSED == CLOSED {
```

Comment:
**if** I'm reading this right, you should be able to read this when "locking" the slot (call to `try_rx_lock)`.

Again, not super critical at this point, but a comment saying "we _could_ do this" could be helpful for anyone working in this file later.

Reply:
Good call. I moved this to where we set the closed bit or write a new value

Reply:
I've added to the comment above. I have an idea for how this could happen, but it will require some additional changes to what `try_rx_lock` returns and how we properly spin when it's locked by a writer.

I started making the change as part of this, but will keep it separate since it is not really related to this fix.